File: src/main/scala/com/vpn/mine/activity/MainActivity.java
package com.vpn.mine.activity;

import android.os.Handler;
import android.os.IBinder;
import android.os.RemoteException;
import android.support.design.widget.TabLayout;
import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentManager;
import android.support.v4.view.PagerAdapter;
import android.support.v4.view.ViewPager;
import android.os.Bundle;
import android.support.v7.app.AppCompatActivity;
import android.view.View;
import android.view.ViewGroup;

import com.vpn.mine.Reality;
import com.vpn.mine.ServiceBoundContext;
import com.vpn.mine.activity.ConnectFragment;
import com.vpn.mine.activity.HelpFragment;
import com.vpn.mine.activity.MainFragmentPagerAdapter;
import com.vpn.mine.activity.OptionFragment;
import com.vpn.mine.activity.PurchaseFragment;
import com.vpn.mine.R;
import com.vpn.mine.aidl.IShadowsocksService;
import com.vpn.mine.aidl.IShadowsocksServiceCallback;


import java.util.ArrayList;

/**
 * Created by coder on 17-7-12.
 */
public class MainActivity extends AppCompatActivity implements TabLayout.OnTabSelectedListener{

    //    四个布局
    private View connectView ;
    private View purchaseView;
    private View helpView;
    private View optionView;

    private ViewPager viewPager ;
    private ArrayList<Fragment> fragmentList;

    //  四个标签 和 TabLayout
    private TabLayout tabLayout ;
    private TabLayout.Tab connectTab;
    private TabLayout.Tab purchaseTab;
    private TabLayout.Tab helpTab;
    private TabLayout.Tab optionTab;

    private ConnectFragment connectFragment;
    private PurchaseFragment purchaseFragment;
    private HelpFragment helpFragment;
    private OptionFragment optionFragment;

    private Handler handler = new Handler();

    private IShadowsocksServiceCallback.Stub callback = new IShadowsocksServiceCallback.Stub() {
        @Override
        public void stateChanged(final int state, String profileName, String msg) throws RemoteException {
            boolean post = handler.post(new Runnable() {
                @Override
                public void run() {
                    switch (state) {
                        case 1://connecting
//                            调整按钮
                            break;
                        case 2://connected

                            break;
                        case 4://stoped

                            break;

                    }
                }
            });
        }

        @Override
        public void trafficUpdated(long txRate, long rxRate, long txTotal, long rxTotal) throws RemoteException {

        }
    };


    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.index);

        //创建控件和点击事件
        initView();
        initEvents();

        //调用attachService
        handler.post(new Runnable() {
            @Override
            public void run() {
                new Reality().attachService(callback);
            }
        });
    }

    private void initEvents() {
        tabLayout.addOnTabSelectedListener(new TabLayout.OnTabSelectedListener(){

            @Override
            public void onTabSelected(TabLayout.Tab tab) {
                if (tab == tabLayout.getTabAt(0)){
                    viewPager.setCurrentItem(0);
                } else if (tab == tabLayout.getTabAt(1)){
                    viewPager.setCurrentItem(1);
                } else if (tab == tabLayout.getTabAt(2)){
                    viewPager.setCurrentItem(2);
                } else if (tab == tabLayout.getTabAt(3)){
                    viewPager.setCurrentItem(3);
                }
            }

            @Override
            public void onTabUnselected(TabLayout.Tab tab) {

            }

            @Override
            public void onTabReselected(TabLayout.Tab tab) {

            }
        });
    }

    private void initView() {
        connectFragment  = new ConnectFragment();
        purchaseFragment = new PurchaseFragment();
        helpFragment = new HelpFragment();
        optionFragment = new OptionFragment();
        tabLayout = (TabLayout) findViewById(R.id.tabLayout);
        viewPager = (ViewPager) findViewById(R.id.viewpager);

        fragmentList = new ArrayList<Fragment>();
        fragmentList.add(connectFragment);
        fragmentList.add(purchaseFragment);
        fragmentList.add(helpFragment);
        fragmentList.add(optionFragment);

        FragmentManager fragmentManager = this.getSupportFragmentManager();
        viewPager.setAdapter(new MainFragmentPagerAdapter(fragmentManager,fragmentList));
        //默认显示连接页面
        viewPager.setCurrentItem(0);

        //设置tabLayout与viewPager级联
        tabLayout.setupWithViewPager(viewPager);

        //实例化Tab
        connectTab = tabLayout.getTabAt(0);
        purchaseTab = tabLayout.getTabAt(1);
        helpTab = tabLayout.getTabAt(2);
        optionTab = tabLayout.getTabAt(3);

        //重新设置标签名称，上面会清除全部标签
        connectTab.setText("连接");
        purchaseTab.setText("购买");
        helpTab.setText("帮助");
        optionTab.setText("选项");
    }

//  以下三个方法是Tab事件

    @Override
    public void onTabSelected(TabLayout.Tab tab) {

    }

    @Override
    public void onTabUnselected(TabLayout.Tab tab) {

    }

    @Override
    public void onTabReselected(TabLayout.Tab tab) {

    }





    class NewPagerAdapter extends PagerAdapter{

        private ArrayList<View> viewList;

        public NewPagerAdapter(ArrayList<View> viewList) {
            this.viewList = viewList;
        }

        @Override
        public int getCount() {
            return viewList.size();
        }

        @Override
        public boolean isViewFromObject(View view, Object object) {
            return view == object;
        }

        @Override
        public void destroyItem(ViewGroup container, int position, Object object) {
            container.removeView(viewList.get(position));
        }

        @Override
        public int getItemPosition(Object object) {
            return super.getItemPosition(object);
        }

        @Override
        public Object instantiateItem(ViewGroup container, int position) {
            container.addView(viewList.get(position));
            return viewList.get(position);
        }
    }

    class MyOnClickListener implements View.OnClickListener {

        private int index;

        public MyOnClickListener(int index) {
            this.index = index;
        }

        @Override
        public void onClick(View view) {
            viewPager.setCurrentItem(index);
        }
    }
}

File: src/main/scala/com/vpn/mine/activity/MainActivity.scala
package com.vpn.mine.activity

import java.util
import java.util.ArrayList

import android.app.Activity
import android.net.VpnService
import android.os.{Bundle, Handler}
import android.support.design.widget.TabLayout
import android.support.v4.app.{Fragment, FragmentManager}
import android.support.v4.view.{PagerAdapter, ViewPager}
import android.support.v7.app.AppCompatActivity
import android.view.{View, ViewGroup}
import com.vpn.mine.aidl.IShadowsocksServiceCallback
import com.vpn.mine.job.SSRSubUpdateJob
import com.vpn.mine.utils.{DataSaver, State}
import com.vpn.mine.{R, ServiceBoundContext}
import com.vpn.mine.MyApplication.app
/**
  * Created by coder on 17-7-14.
  */
class MainActivity extends AppCompatActivity with TabLayout.OnTabSelectedListener with ServiceBoundContext{


  //    四个布局
  private val connectView: View = null
  private val purchaseView: View = null
  private val helpView: View = null
  private val optionView: View = null

  private var viewPager: ViewPager = null
  private var fragmentList: util.ArrayList[Fragment] = null

  //  四个标签 和 TabLayout
  private var tabLayout: TabLayout = null
  private var connectTab: TabLayout.Tab = null
  private var purchaseTab: TabLayout.Tab = null
  private var helpTab: TabLayout.Tab = null
  private var optionTab: TabLayout.Tab = null

  private var connectFragment: ConnectFragment = null
  private var purchaseFragment: PurchaseFragment = null
  private var helpFragment: HelpFragment = null
  private var optionFragment: OptionFragment = null

  private val handler = new Handler

  private val callBack : IShadowsocksServiceCallback.Stub = new IShadowsocksServiceCallback.Stub {

    override def stateChanged(s: Int, profileName: String, m: String): Unit = {
      handler.post(() => {
        s match {
          case State.CONNECTING =>
            //正在连接
            println("正在连接")
          case State.CONNECTED =>
            //已连接
            println("已连接")
          case State.STOPPED =>
            //已停止
            println("已停止")
          case State.STOPPING =>
            //正在停止
            println("正在停止")
        }
      })
    }

    override def trafficUpdated(txRate: Long, rxRate: Long, txTotal: Long, rxTotal: Long): Unit = Unit

  }


  override def onCreate(savedInstanceState: Bundle): Unit = {
    super.onCreate(savedInstanceState)
    setContentView(R.layout.index)

    //创建控件和点击事件
    initView()
    initEvent()


    println("执行有关数据库的操作")
    //和数据库有关的操作
    app.ssrsubManager.getFirstSSRSub match {
      case Some(first) => {

      }
      case None => app.ssrsubManager.createDefault()
    }

    println("接下来执行schedule")
    SSRSubUpdateJob.schedule()

    println("调用attachService")
    //调用attachService
    handler.post(() => attachService(callBack)
    )
  }
  protected override def onResume() {
    super.onResume()

    app.refreshContainerHolder

    //待添加更新当前状态
//    updateState(updateCurrentProfile())
  }

  override def onStart() {
    super.onStart()
    registerCallback
  }

  override def onDestroy() {
    super.onDestroy()
    detachService()
    //待添加
    /*if (receiver != null) {
      unregisterReceiver(receiver)
      receiver = null
    }*/
  }



  def attachService: Unit = attachService(callBack)

  def initView(): Unit ={
    connectFragment = new ConnectFragment
    purchaseFragment = new PurchaseFragment
    helpFragment = new HelpFragment
    optionFragment = new OptionFragment
    tabLayout = findViewById(R.id.tabLayout).asInstanceOf[TabLayout]
    viewPager = findViewById(R.id.viewpager).asInstanceOf[ViewPager]

    fragmentList = new util.ArrayList[Fragment]()
    fragmentList.add(connectFragment)
    fragmentList.add(purchaseFragment)
    fragmentList.add(helpFragment)
    fragmentList.add(optionFragment)

    val fragmentManager : FragmentManager = this.getSupportFragmentManager
    viewPager.setAdapter(new MainFragmentPagerAdapter(fragmentManager, fragmentList))
    //默认显示连接页面
    viewPager.setCurrentItem(0)

    //设置tabLayout与viewPager级联
    tabLayout.setupWithViewPager(viewPager)
    //实例化Tab
    connectTab = tabLayout.getTabAt(0)
    purchaseTab = tabLayout.getTabAt(1)
    helpTab = tabLayout.getTabAt(2)
    optionTab = tabLayout.getTabAt(3)

    //重新设置标签名称，上面会清除全部标签
    connectTab.setText("连接")
    purchaseTab.setText("购买")
    helpTab.setText("帮助")
    optionTab.setText("选项")

  }


  def initEvent(): Unit ={
    tabLayout.addOnTabSelectedListener(new TabLayout.OnTabSelectedListener {

      override def onTabSelected(tab: TabLayout.Tab): Unit = {
        if (tab eq tabLayout.getTabAt(0)) viewPager.setCurrentItem(0)
        else if (tab eq tabLayout.getTabAt(1)) viewPager.setCurrentItem(1)
        else if (tab eq tabLayout.getTabAt(2)) viewPager.setCurrentItem(2)
        else if (tab eq tabLayout.getTabAt(3)) viewPager.setCurrentItem(3)
      }

      override def onTabUnselected(tab: TabLayout.Tab): Unit = Unit

      override def onTabReselected(tab: TabLayout.Tab): Unit = Unit
    })
  }

  override def onTabSelected(tab: TabLayout.Tab): Unit = Unit

  override def onTabReselected(tab: TabLayout.Tab): Unit = Unit

  override def onTabUnselected(tab: TabLayout.Tab): Unit = Unit

  class NewPagerAdapter(views : ArrayList[View]) extends PagerAdapter{

    var viewList :ArrayList[View] = _

    this.viewList = views

    override def getCount: Int = viewList.size

    override def isViewFromObject(view: View, o: scala.Any): Boolean = view == o

    override def destroyItem(container: ViewGroup, position: Int, `object`: scala.Any): Unit = {
      container.removeView(viewList.get(position))
    }

    override def getItemPosition(`object`: scala.Any): Int = {
      super.getItemPosition(`object`)
    }

    override def instantiateItem(container: ViewGroup, position: Int): AnyRef = {
      container.addView(viewList.get(position))
      viewList.get(position)
    }
  }

  def prepareStartService(): Unit ={
//    println("进来了")
    //连接VPn
    val vpnIntent = VpnService.prepare(this)
    //如果当前系统中没有VPN连接，或者存在的VPN连接不是本程序建立的
    //则VpnService.prepare函数会返回一个intent。这个intent就是用来触发确认对话框的
    if (vpnIntent != null) startActivityForResult(vpnIntent, 0)
    else { //如果当前系统中有VPN连接，并且这个连接就是本程序建立的，则函数会返回null，就不需要用户再确认了
      onActivityResult(0, Activity.RESULT_OK, null)
    }
  }

  def serviceLoad(): Unit ={
    //更新profileId的值
    app.profileId(DataSaver.NODE_INDEX)
//    println(bgService != null)
    bgService.use(app.profileId)
    println("use方法调用结束")
    //待添加改变连接的状态显示

  }
}
